### PR TITLE
[ENH] Pure Nipype brain extraction workflow

### PR DESCRIPTION
**Pull request by oesteban**
Using the implementation of ``antsBrainExtraction.sh`` from niworkflows
(supersedes #48)

### CODE REVIEW

File: smriprep/workflows/anatomical.py [hunk 1]
```diff
@@ -209,11 +209,18 @@ def init_anat_preproc_wf(skull_strip_template, fs_spaces, template, debug,
 [@ants, RRID:SCR_004757]"""
     desc += '.\n' if num_t1w > 1 else ", and used as T1w-reference throughout the workflow.\n"
 
+    desc = """\
+The T1w-reference was then skull-stripped with a *Nipype* implementation of
+the `antsBrainExtraction.sh` workflow (from ANTs), using {skullstrip_tpl}
+as target template.
+""".format(skullstrip_tpl=skull_strip_template)
```
Comment (effigies) [inline smriprep/workflows/anatomical.py]: What's our minimum Python? If it's 3.6 or higher, we can switch to [f-strings](https://docs.python.org/3/reference/lexical_analysis.html#formatted-string-literals).

File: smriprep/workflows/anatomical.py [hunk 2]
```diff
@@ -231,15 +241,18 @@ def init_anat_preproc_wf(skull_strip_template, fs_spaces, template, debug,
 
     # 3. Skull-stripping
     # Bias field correction is handled in skull strip workflows.
-    skullstrip_ants_wf = init_skullstrip_ants_wf(name='skullstrip_ants_wf',
-                                                 skull_strip_template=skull_strip_template,
-                                                 debug=debug,
-                                                 omp_nthreads=omp_nthreads)
+    bran_extraction_wf = init_brain_extraction_wf(
```
Comment (effigies) [inline smriprep/workflows/anatomical.py]:
```suggestion
    brain_extraction_wf = init_brain_extraction_wf(
```

File: smriprep/workflows/anatomical.py (hunk 1 above)
Comment (oesteban) [reply smriprep/workflows/anatomical.py]: I was actually thinking of that as well. We have Python +3.5 but I'd seriously consider enforcing +3.6 for that very reason.

File: smriprep/workflows/anatomical.py (hunk 2 above)
Comment (oesteban) [reply smriprep/workflows/anatomical.py]: OMG never push a commit while you hear your kid crying :(

**Comment by oesteban** [comment]
Before: https://256-160736148-gh.circle-artifacts.com/0/tmp/ds054/derivatives/smriprep/sub-100185.html
After: https://213-164755300-gh.circle-artifacts.com/0/tmp/ds054/derivatives/smriprep/sub-100185.html

**Comment by effigies** [comment]
> Note: none of the builds are executed with fixed seed for atropos, hence the little differences.

Is it possible to fix the seed by environment variable?

**Comment by oesteban** [comment]
We can include that. Are you thinking of a shared env variable for all algorithms accepting a seed?

Something like `$BIDS_PREP_SEED` ?

**Comment by effigies** [comment]
No, I meant more "Does Atropos accept a seed"? I would rather we let these things be controlled by the underlying tools if at all possible, just to keep things simple.

**Comment by effigies** [comment]
I suppose I should say this shouldn't hold up this PR. I think we should be deliberate and perhaps solicit input from @repronim before settling on an approach, anyway. (Also, it looks like ANTs does have some work in this direction: https://github.com/ANTsX/ANTs/pull/599)

**Comment by oesteban** [comment]
[Atropos does not accept a seed, it only provides a boolean switch to make it random or not](https://github.com/ANTsX/ANTs/blob/15f4e4013ed33b9d226c8d1b7d9509b2a8b19ba2/Examples/Atropos.cxx#L760)

**Comment by effigies** [comment]
I wonder if we can persuade them to move in a more consistent direction with the Registration stuff.

**Comment by oesteban** [comment]
Alright, is it okay to merge then?

File: smriprep/workflows/anatomical.py [hunk 3]
```diff
@@ -258,12 +270,12 @@ def init_anat_preproc_wf(skull_strip_template, fs_spaces, template, debug,
...
                 ('subject_id', 'inputnode.subject_id')]),
             (anat_template_wf, surface_recon_wf, [('outputnode.t1_template', 'inputnode.t1w')]),
-            (skullstrip_ants_wf, surface_recon_wf, [
-                ('outputnode.out_file', 'inputnode.skullstripped_t1'),
-                ('outputnode.out_segs', 'inputnode.ants_segs'),
-                ('outputnode.bias_corrected', 'inputnode.corrected_t1')]),
-            (skullstrip_ants_wf, applyrefined, [
-                ('outputnode.bias_corrected', 'in_file')]),
+            (brain_extraction_wf, surface_recon_wf, [
+                (('outputnode.bias_corrected', _pop), 'inputnode.skullstripped_t1'),
```
Comment (effigies) [inline smriprep/workflows/anatomical.py]: Bias corrected isn't the same as skull stripped, is it?

Comment (oesteban) [reply smriprep/workflows/anatomical.py]: Yes, it was. This is one of the reasons why I wanted to translate this workflow into nipype. In a future PR I want to address these:

- Bias-corrected should not be the same as skull stripped
- Add an additional N4 pass on the original image (not truncated, as it is now) with the brain mask, so both bias-corrected and skull-stripped have this more accurate correction.
- Reuse the initial transforms in the normalization workflow if the template used is the same (and consequently move to MNI152NLin2009cAsym as target template for skull-stripping).

Comment (effigies) [reply smriprep/workflows/anatomical.py]: But we save the bias-corrected file as the t1_preproc (L253), and that's not skull-stripped.

Comment (oesteban) [reply smriprep/workflows/anatomical.py]: You are right. It should be fixed now.

**Pull request merged**